tutorials/14-minituring/14_minituring.jmd

```diff
@@ -0,0 +1,300 @@
+end
+
+function _assume(__context__::SamplingContext, __varinfo__, sampler, dist, var_id)
+    sample = Random.rand(__context__.rng, dist)
+    log_p = Distributions.loglikelihood(dist, sample)
+    update_varinfo!(__varinfo__, var_id, sample, log_p)
+    return sample
+end
+```
+
+Then, we'll encounter our first nontrivial piece of Julia code, that is the compiler. In one sentence, the compiler will replace every tilde notation with a call to `observe` or `assume` function defined in the `context` code block.
+
+> **Julia Sidebar**  
+I am afraid there is not easy way to explain what happens here. So I am going leave some references, and once the reader go though the referenced material, the following block should not be too difficult to understand.  
+[Julia Metaprogramming](https://docs.julialang.org/en/v1/manual/metaprogramming/)  
+[MacroTools](https://fluxml.ai/MacroTools.jl/dev/pattern-matching/)  
+[Turing Compiler Design](https://turing.ml/dev/docs/for-developers/compiler) (The implementation here is heavily simplified compared to Turing's compiler.)
+
+```julia
+macro mini_model(expr)
+    return esc(_mini_model(expr, __source__, __module__))
+end
+
+function _mini_model(expr, __source__, __module__)
+    _def = MacroTools.splitdef(expr)
+    _def_args, _def_body = _def[:args], _def[:body]
+    
+    _def_body = MacroTools.postwalk(_def_body) do sub_expr
+        # MacroTools' macro `capture` provide handy utility for pattern matching
+        if MacroTools.@capture(sub_expr, var_ ~ dist_)
+            if var in _def_args
+                # MacroTools' macro `q` gives expr without linenum
+                return MacroTools.@q begin
+                    $(Main._observe)(__context__, __varinfo__, __model__.sampler, $dist, $(Meta.quot(var)), $var)
+                end
+            else
+                return MacroTools.@q begin
+                    $var = $(Main._assume)(__context__, __varinfo__, __model__.sampler, $dist, $(Meta.quot(var)))
+                end
+            end
+        else 
+            return sub_expr
+        end
+    end
+    
+    _def[:args] = vcat(
+        [
+            # Insert extra arguments
+            :(__model__::($Main.MiniModel)),
+            :(__varinfo__::($Main.AbstractVarInfo)),
+            :(__context__::($Main.AbstractContext)),
+        ],
+        _def_args,
+    )
+    
+    _def[:body] = MacroTools.@q begin
+        $__source__
+        $_def_body
+    end
+    
+    return MacroTools.@q begin
+        $(MacroTools.combinedef(_def))
+    end
+end
+```
+
+Next, let's define a `Model` struct. In the actual implementation of the compiler in Turing, the model construction is done as the final step of the compilation. But for the sake of simplicity, we will construct the model manually. This also requires us to keep a field for the actually data in the `Model` struct. 
+
+```julia
+struct MiniModel<:AbstractMCMC.AbstractModel
+    f
+    sampler
```

Comment:
The `sampler` is not part of the model in Turing, and also not in the AbstractMCMC interface, but provided separately to `sample`. Remove it here?
```suggestion
```